### PR TITLE
Run plugins inside Docker, not Firecracker.

### PR DESCRIPTION
**Pull request by wimax-grapl**
### Which issue does this PR correspond to?
https://github.com/grapl-security/issue-tracker/issues/908

### What changes does this PR make to Grapl? Why?
- Swap out the job we run for plugins. It's now a debian slim that executes a mounted artifact.
- Notably, and probably to some peoples' chagrin, I am keeping the firecracker job in there (and tested) to ensure it doesn't regress.

### How were these changes tested?
I shoved in a "real" binary that's just a small ELF hello world.
Looked at the logs and, shock, it works!
```
ubuntu@ip-172-31-32-238:[...]/plugin-a972930a-6ced-d4c2-b112-0efebe366a8f/grapl-plugin$ 
cat run-plugin.stdout.ab29283b.log 
Hi World

ubuntu@ip-172-31-32-238:[...]/plugin-a972930a-6ced-d4c2-b112-0efebe366a8f/grapl-plugin$ 
cat run-plugin.events.ab29283b.log 
Task received by client
Building Task Directory
Client is downloading artifacts
Task started by client
Exit Code: 0
Task restarting in 16.134713417s
```

### CODE REVIEW

File: src/rust/plugin-registry/src/server/deploy_plugin.rs [hunk 1]
```diff
@@ -28,6 +29,83 @@ use crate::{
...
+                (
+                    "plugin_runtime_image",
+                    service_config.hax_docker_plugin_runtime_image.to_owned(),
+                ),
+                ("plugin_id", plugin.plugin_id.to_string()),
+                ("tenant_id", plugin.tenant_id.to_string()),
+            ]);
+            cli.parse_hcl2(job_file_hcl, job_file_vars)
+        }
+        PluginRuntime::Firecracker => {
```
Comment (wimax-grapl) [inline src/rust/plugin-registry/src/server/deploy_plugin.rs:75]: this is the stuff that we previously had on line 42 on the red side

File: nomad/grapl-core.nomad [hunk 2]
```diff
@@ -1354,16 +1354,17 @@ job "grapl-core" {
...
+        NOMAD_SERVICE_ADDRESS                           = "${attr.unique.network.ip-address}:4646"
+        PLUGIN_REGISTRY_BIND_ADDRESS                    = "0.0.0.0:${NOMAD_PORT_plugin-registry-port}"
+        PLUGIN_REGISTRY_DB_HOSTNAME                     = var.plugin_registry_db_hostname
+        PLUGIN_REGISTRY_DB_PASSWORD                     = var.plugin_registry_db_password
+        PLUGIN_REGISTRY_DB_PORT                         = var.plugin_registry_db_port
+        PLUGIN_REGISTRY_DB_USERNAME                     = var.plugin_registry_db_username
+        PLUGIN_BOOTSTRAP_CONTAINER_IMAGE                = var.container_images["plugin-bootstrap"]
+        PLUGIN_REGISTRY_KERNEL_ARTIFACT_URL             = var.plugin_registry_kernel_artifact_url
+        PLUGIN_REGISTRY_ROOTFS_ARTIFACT_URL             = var.plugin_registry_rootfs_artifact_url
+        PLUGIN_REGISTRY_HAX_DOCKER_PLUGIN_RUNTIME_IMAGE = var.container_images["hax-docker-plugin-runtime"]
```
Comment (wimax-grapl) [inline nomad/grapl-core.nomad:1367]: this 'hax' is carrying over a pattern commonly used at my previous company for basically blatantly signaling to everybody who sees a piece of code that "this is going away soon and is not necessarily high quality, by intention"

File: pulumi/grapl/__main__.py [hunk 3]
```diff
@@ -59,6 +59,7 @@ def _container_images(artifacts: ArtifactGetter) -> Mapping[str, DockerImageId]:
         "engagement-creator": builder.build_with_tag("engagement-creator"),
         "graph-merger": builder.build_with_tag("graph-merger"),
         "graphql-endpoint": builder.build_with_tag("graphql-endpoint"),
+        "hax-docker-plugin-runtime": DockerImageId("debian:bullseye-slim"),
```
Comment (wimax-grapl) [inline pulumi/grapl/__main__.py:62]: "sure, why not"

Comment (wimax-grapl) [reply pulumi/grapl/__main__.py:62]: maybe alpine is better, we want a pretty minimal image

Comment (twunderlich-grapl) [reply pulumi/grapl/__main__.py:62]: I think bullseye-slim is fine for now. Alpine can have issues with C dependencies, which can be problematic with certain python wheels

File: src/rust/Dockerfile [hunk 4]
```diff
@@ -28,15 +28,17 @@ RUN --mount=type=cache,target=/var/lib/apt/lists,sharing=locked,id=rust-base-apt
...
 RUN <<EOF
-set -euo pipefail
+NOMAD_VERSION="1.2.4"
+ZIP_NAME="nomad_${NOMAD_VERSION}_linux_amd64.zip"
 curl --remote-name --proto '=https' --tlsv1.2 -sSf \
-  https://releases.hashicorp.com/nomad/1.2.4/nomad_1.2.4_linux_amd64.zip
-unzip *.zip
-rm *.zip
+  "https://releases.hashicorp.com/nomad/${NOMAD_VERSION}/${ZIP_NAME}"
+unzip "${ZIP_NAME}"
```
Comment (twunderlich-grapl) [inline src/rust/Dockerfile:40]: much nicer

File: src/rust/plugin-registry/src/static_files/hax_docker_plugin.nomad [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+job "grapl-plugin" {
+  datacenters = ["dc1"]
+  namespace   = "plugin-${var.plugin_id}"
+  type        = "service"
+
+  # We'll want to make sure we have the opposite constraint on other services
+  # This is set in the Nomad agent's `client` stanza:
+  # https://www.nomadproject.io/docs/configuration/client#meta
+  constraint {
+    attribute = "${meta.is_grapl_plugin_host}"
```
Comment (twunderlich-grapl) [inline src/rust/plugin-registry/src/static_files/hax_docker_plugin.nomad:41]: Are you thinking that we'll have one cluster for plugins or that we'll have tenant-specific clusters?
Also I think `meta.is_grapl_plugin_host` would work here

**Review by twunderlich-grapl — APPROVED** [review]
Approving the non-rust code.

File: src/rust/plugin-registry/src/static_files/hax_docker_plugin.nomad (hunk 5 above)
Comment (wimax-grapl) [reply src/rust/plugin-registry/src/static_files/hax_docker_plugin.nomad:41]: This idea is originally from the RFC which I blindly followed without thinking about it:
https://github.com/grapl-security/grapl-rfcs/blob/main/text/0000-plugins.md#nomad-job

But it's actually really interesting if we, say, had some Nomad agents that ran on a1 metal instances just for plugins, and other ones that ran on other types for regular services, for example

File: pulumi/grapl/__main__.py (hunk 3 above)
Comment (colin-grapl) [reply pulumi/grapl/__main__.py:62]: Agree that debian-slim is fine

File: src/rust/plugin-registry/src/static_files/hax_docker_plugin.nomad (hunk 5 above)
Comment (colin-grapl) [reply src/rust/plugin-registry/src/static_files/hax_docker_plugin.nomad:41]: The idea is to colocate tenants on the AWS hardware, which is why we rely so heavily on Firecracker.

**Review by colin-grapl — APPROVED** [review]

**Pull request merged**